### ISSUE TITLE
switch to pyzmq?

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/pescador#6]
IPC via multiprocessing is unbearably slow for me; it overshadows any benefits due to forking the data io into another thread.

@ejhumphrey Any thoughts about trying pyzmq instead?  I'm not sure how much think work needs to go into it, but it seems promising.

**Comment by ejhumphrey** [comment]
based on dozens of seconds of skimming: promising, yes; think work,
potentially large.

I'd push it on the stack for post-dissertation inquiries, but don't let my
queue hold you back.

**Comment by bmcfee** [comment]
I might take a crack at it this week.  The main benefit here would be shared buffer access to ndarrays, so we can bypass that whole batch serialization issue.  See: http://zeromq.github.io/pyzmq/serialization.html

**Comment by bmcfee** [comment]
Dropping this here also: https://learning-0mq-with-pyzmq.readthedocs.org/en/latest/index.html

specifically: https://learning-0mq-with-pyzmq.readthedocs.org/en/latest/pyzmq/multiprocess/multiprocess.html and https://learning-0mq-with-pyzmq.readthedocs.org/en/latest/pyzmq/multisocket/zmqpoller.html

It took about four seconds to get a demo of shared ndarray buffers working with 0mq and multiprocessing: https://gist.github.com/e26a8d160ede121d54e9

Bundling that up inside a pescador generator, and having the reader use zmqpoller should pretty much solve everything.

**Comment by ejhumphrey** [comment]
>
> pretty much solve everything.
>
​no sweeter words in the English language.​

**Comment by bmcfee** [comment]
And here's a more realistic demo using multi-part messages to send ditcs of ndarrays over the pipe.  https://gist.github.com/bmcfee/98eff9b7575d18bdb039

**Comment by bmcfee** [comment]
One potential downside here: scipy.sparse matrices don't implement the buffer protocol, so we can only ship dense data over zmq.

**Comment by bmcfee** [comment]
5509343e9036233a6cd54af2f70eaa0a4b7304a3 bam.

On my simple 5-fold cross validation, iris+sgdclassifier+random noise augmentation benchmark, I see the following results with a batch size of 16:

Single thread
```
CPU times: user 4.25 s, sys: 7.32 ms, total: 4.26 s
Wall time: 4.25 s
```

threaded_mux
```
CPU times: user 9.13 s, sys: 394 ms, total: 9.52 s
Wall time: 9.52 s
```

zmq_mux
```
CPU times: user 2.66 s, sys: 460 ms, total: 3.12 s
Wall time: 4.92 s
```

This can probably be accelerated further by pushing a buffer into the worker thread to cut down on communication.

TLDR: worth it!

**Comment by ejhumphrey** [comment]
halp me not be dumb. wall clocks r same? is the data throughput higher on
the zmq_mux?

**Comment by bmcfee** [comment]
wall clocks are comparable for single vs zmq on this data for a few reasons:

* everything's in memory already, so the streams aren't disk-bound
* the data's very low dimensional

this benchmark is just to show that we can do multiprocessing without high communication overhead.

**Comment by ejhumphrey** [comment]
derp kthxbai

**Comment by bmcfee** [comment]
I just redid the benchmarks so that the data generator pads with `2000 - d` dimensions of random gaussian noise. Results:

Single
```
Test-set accuracy: 0.967
Test-set accuracy: 0.867
Test-set accuracy: 0.933
Test-set accuracy: 0.900
Test-set accuracy: 0.933
CPU times: user 17.7 s, sys: 62.3 ms, total: 17.7 s
Wall time: 17.7 s
```

Multiprocessing.Queue
```
Test-set accuracy: 0.933
Test-set accuracy: 1.000
Test-set accuracy: 0.900
Test-set accuracy: 0.967
Test-set accuracy: 0.900
CPU times: user 21.5 s, sys: 739 ms, total: 22.2 s
Wall time: 22.2 s
```

ZMQ
```
Test-set accuracy: 0.933
Test-set accuracy: 1.000
Test-set accuracy: 0.900
Test-set accuracy: 0.967
Test-set accuracy: 0.900
CPU times: user 12.6 s, sys: 487 ms, total: 13.1 s
Wall time: 12.4 s
```

I think we can stick a fork in this one, but i'm opening a new issue for a related problem...

**Comment by urinieto** [comment]
ALRIGHTO.
